File: app/src/views/Humans/morphing/dual-contouring.ts
// Dual Contouring surface extraction (Surface Nets variant)
// Based on: https://people.compute.dtu.dk/janba/gallery/polygonization.html
//
// STATUS: Basic implementation complete - produces closed meshes with gradient-based
// vertex projection. Currently using simple 2x2 grid face generation.
//
// TODO for better quality:
// - Fix triangle winding order for proper front-face culling (currently using DoubleSide)
// - Implement proper QEF (Quadratic Error Function) vertex positioning
// - Add edge-based quad generation instead of grid-based
// - Consider full marching cubes lookup table for better topology

import type { VoxelGrid } from "./voxel-grid";
import type { Vec3 } from "./types";
import { computeGradient } from "./sdf";

type MeshData = {
  vertices: number[];
  indices: number[];
};

type CellVertex = {
  position: Vec3;
  index: number;
};

/**
 * Extract triangle mesh using Dual Contouring
 * Produces higher quality meshes than Marching Cubes by:
 * - Placing one vertex per cell (not per edge)
 * - Projecting vertices onto the actual isosurface
 * - Generating quads that triangulate cleanly
 */
export function dualContouring(
  grid: VoxelGrid,
  evaluateSDF: (p: Vec3) => number,
  isoValue: number = 0
): MeshData {
  const vertices: number[] = [];
  const indices: number[] = [];
  const res = grid.resolution;

  // Store vertex index for each cell that contains the surface
  const cellVertices = new Map<string, CellVertex>();

  // Step 1: Create vertices for cells that intersect the isosurface
  for (let z = 0; z < res - 1; z++) {
    for (let y = 0; y < res - 1; y++) {
      for (let x = 0; x < res - 1; x++) {
        const cellKey = `${x},${y},${z}`;

        // Check if this cell intersects the isosurface
        if (!cellIntersectsSurface(grid, x, y, z, isoValue)) {
          continue;
        }

        // Find best vertex position for this cell
        const vertexPos = findCellVertex(grid, evaluateSDF, x, y, z, isoValue);

        // Add to vertex array
        const vertexIndex = vertices.length / 3;
        vertices.push(vertexPos[0], vertexPos[1], vertexPos[2]);

        cellVertices.set(cellKey, {
          position: vertexPos,
          index: vertexIndex,
        });
      }
    }
  }

  // Step 2: Generate faces between adjacent cells
  // Connect cells that both have vertices and share a face
  for (let z = 0; z < res - 1; z++) {
    for (let y = 0; y < res - 1; y++) {
      for (let x = 0; x < res - 1; x++) {
        const cell = cellVertices.get(`${x},${y},${z}`);
        if (!cell) continue;

        // Create face in +X direction (YZ plane)
        if (x < res - 2) {
          createFaceX(cellVertices, indices, x, y, z);
        }

        // Create face in +Y direction (XZ plane)
        if (y < res - 2) {
          createFaceY(cellVertices, indices, x, y, z);
        }

        // Create face in +Z direction (XY plane)
        if (z < res - 2) {
          createFaceZ(cellVertices, indices, x, y, z);
        }
      }
    }
  }

  return { vertices, indices };
}

/**
 * Check if a voxel cell intersects the isosurface
 * Returns true if corners have different signs (sign change = surface crossing)
 */
function cellIntersectsSurface(
  grid: VoxelGrid,
  x: number,
  y: number,
  z: number,
  isoValue: number
): boolean {
  // Get 8 corner values
  const corners = [
    grid.get(x, y, z),
    grid.get(x + 1, y, z),
    grid.get(x + 1, y, z + 1),
    grid.get(x, y, z + 1),
    grid.get(x, y + 1, z),
    grid.get(x + 1, y + 1, z),
    grid.get(x + 1, y + 1, z + 1),
    grid.get(x, y + 1, z + 1),
  ];

  // Check if any corner is inside (< isoValue) and any is outside (>= isoValue)
  let hasInside = false;
  let hasOutside = false;

  for (const value of corners) {
    if (value < isoValue) {
      hasInside = true;
    } else {
      hasOutside = true;
    }
  }

  return hasInside && hasOutside;
}

/**
 * Find optimal vertex position for a cell using surface projection
 * Uses gradient descent to project cell center onto isosurface
 */
function findCellVertex(
  grid: VoxelGrid,
  evaluateSDF: (p: Vec3) => number,
  x: number,
  y: number,
  z: number,
  isoValue: number
): Vec3 {
  // Start at cell center
  const cellCenter = grid.getPosition(x + 0.5, y + 0.5, z + 0.5);
  let pos: Vec3 = [cellCenter[0], cellCenter[1], cellCenter[2]];

  // Simple iterative projection: move along gradient toward surface
  const maxIterations = 8;
  for (let iter = 0; iter < maxIterations; iter++) {
    const dist = evaluateSDF(pos) - isoValue;

    // Close enough to surface
    if (Math.abs(dist) < 0.001) {
      break;
    }

    // Compute gradient at current position
    const grad = computeGradient(pos, evaluateSDF);

    // Move toward surface (gradient points away from surface, so subtract)
    const stepSize = dist * 0.5; // Damped step
    pos = [
      pos[0] - grad[0] * stepSize,
      pos[1] - grad[1] * stepSize,
      pos[2] - grad[2] * stepSize,
    ];
  }

  return pos;
}

/**
 * Create face between cells in X direction
 * Connects 4 cells in a 2x2 grid in the YZ plane
 */
function createFaceX(
  cellVertices: Map<string, CellVertex>,
  indices: number[],
  x: number,
  y: number,
  z: number
): void {
  // Four cells in YZ plane: (x,y,z), (x,y+1,z), (x,y,z+1), (x,y+1,z+1)
  const v0 = cellVertices.get(`${x},${y},${z}`);
  const v1 = cellVertices.get(`${x},${y + 1},${z}`);
  const v2 = cellVertices.get(`${x},${y + 1},${z + 1}`);
  const v3 = cellVertices.get(`${x},${y},${z + 1}`);

  if (!v0 || !v1 || !v2 || !v3) return;

  // Triangulate quad along shortest diagonal
  const diag02 = distance(v0.position, v2.position);
  const diag13 = distance(v1.position, v3.position);

  if (diag02 < diag13) {
    indices.push(v0.index, v1.index, v2.index);
    indices.push(v0.index, v2.index, v3.index);
  } else {
    indices.push(v0.index, v1.index, v3.index);
    indices.push(v1.index, v2.index, v3.index);
  }
}

/**
 * Create face between cells in Y direction
 * Connects 4 cells in a 2x2 grid in the XZ plane
 */
function createFaceY(
  cellVertices: Map<string, CellVertex>,
  indices: number[],
  x: number,
  y: number,
  z: number
): void {
  // Four cells in XZ plane: (x,y,z), (x+1,y,z), (x,y,z+1), (x+1,y,z+1)
  const v0 = cellVertices.get(`${x},${y},${z}`);
  const v1 = cellVertices.get(`${x + 1},${y},${z}`);
  const v2 = cellVertices.get(`${x + 1},${y},${z + 1}`);
  const v3 = cellVertices.get(`${x},${y},${z + 1}`);

  if (!v0 || !v1 || !v2 || !v3) return;

  const diag02 = distance(v0.position, v2.position);
  const diag13 = distance(v1.position, v3.position);

  if (diag02 < diag13) {
    indices.push(v0.index, v1.index, v2.index);
    indices.push(v0.index, v2.index, v3.index);
  } else {
    indices.push(v0.index, v1.index, v3.index);
    indices.push(v1.index, v2.index, v3.index);
  }
}

/**
 * Create face between cells in Z direction
 * Connects 4 cells in a 2x2 grid in the XY plane
 */
function createFaceZ(
  cellVertices: Map<string, CellVertex>,
  indices: number[],
  x: number,
  y: number,
  z: number
): void {
  // Four cells in XY plane: (x,y,z), (x+1,y,z), (x,y+1,z), (x+1,y+1,z)
  const v0 = cellVertices.get(`${x},${y},${z}`);
  const v1 = cellVertices.get(`${x + 1},${y},${z}`);
  const v2 = cellVertices.get(`${x + 1},${y + 1},${z}`);
  const v3 = cellVertices.get(`${x},${y + 1},${z}`);

  if (!v0 || !v1 || !v2 || !v3) return;

  const diag02 = distance(v0.position, v2.position);
  const diag13 = distance(v1.position, v3.position);

  if (diag02 < diag13) {
    indices.push(v0.index, v1.index, v2.index);
    indices.push(v0.index, v2.index, v3.index);
  } else {
    indices.push(v0.index, v1.index, v3.index);
    indices.push(v1.index, v2.index, v3.index);
  }
}

/**
 * Compute Euclidean distance between two points
 */
function distance(a: Vec3, b: Vec3): number {
  const dx = a[0] - b[0];
  const dy = a[1] - b[1];
  const dz = a[2] - b[2];
  return Math.sqrt(dx * dx + dy * dy + dz * dz);
}

File: app/src-tauri/src/mesh/dual_contouring.rs
use crate::mesh::brick_map::BrickMap;
use crate::mesh::grid_trait::Grid;
use crate::mesh::mould::MouldManager;
use crate::mesh::sdf::compute_gradient;
use crate::mesh::types::{MeshData, Pt3, Vec3};
use crate::mesh::voxel_grid::VoxelGrid;
use rayon::prelude::*;
use std::collections::HashMap;
use std::sync::Mutex;

/// Cell vertex with position and index
#[derive(Debug, Clone)]
struct CellVertex {
    position: Pt3,
    index: u32,
}

/// Extract triangle mesh using Dual Contouring
/// Produces higher quality meshes than Marching Cubes by:
/// - Placing one vertex per cell (not per edge)
/// - Projecting vertices onto the actual isosurface
/// - Generating quads that triangulate cleanly
///
/// Set `fast_mode` to true for interactive previews (skips Newton projection)
pub fn dual_contouring(
    grid: &VoxelGrid,
    mould_manager: &MouldManager,
    iso_value: f32,
) -> MeshData {
    dual_contouring_impl(grid, mould_manager, iso_value, false)
}

/// Fast preview version for realtime interaction
pub fn dual_contouring_fast(
    grid: &VoxelGrid,
    mould_manager: &MouldManager,
    iso_value: f32,
) -> MeshData {
    dual_contouring_impl(grid, mould_manager, iso_value, true)
}

fn dual_contouring_impl(
    grid: &VoxelGrid,
    mould_manager: &MouldManager,
    iso_value: f32,
    fast_mode: bool,
) -> MeshData {
    let res = grid.resolution;

    // Step 1: Create vertices for cells that intersect the isosurface (PARALLEL)
    // Generate all cell coordinates
    let cell_coords: Vec<(u32, u32, u32)> = (0..res - 1)
        .flat_map(|z| {
            (0..res - 1).flat_map(move |y| {
                (0..res - 1).map(move |x| (x, y, z))
            })
        })
        .collect();

    // Parallel vertex creation
    let surface_cells: Vec<((u32, u32, u32), Pt3)> = cell_coords
        .par_iter()
        .filter_map(|&(x, y, z)| {
            // Check if this cell intersects the isosurface
            if !cell_intersects_surface(grid, x, y, z, iso_value) {
                return None;
            }

            // Find best vertex position for this cell
            let vertex_pos = if fast_mode {
                // Fast mode: just use cell center (no Newton iteration)
                grid.get_position(x as f32 + 0.5, y as f32 + 0.5, z as f32 + 0.5)
            } else {
                find_cell_vertex(grid, mould_manager, x, y, z, iso_value)
            };

            Some(((x, y, z), vertex_pos))
        })
        .collect();

    // Build vertices array and cell_vertices map
    let mut vertices: Vec<f32> = Vec::with_capacity(surface_cells.len() * 3);
    let mut cell_vertices: HashMap<(u32, u32, u32), CellVertex> = HashMap::with_capacity(surface_cells.len());

    for ((x, y, z), vertex_pos) in surface_cells {
        let vertex_index = (vertices.len() / 3) as u32;
        vertices.push(vertex_pos.x);
        vertices.push(vertex_pos.y);
        vertices.push(vertex_pos.z);

        cell_vertices.insert(
            (x, y, z),
            CellVertex {
                position: vertex_pos,
                index: vertex_index,
            },
        );
    }

    // Step 2: Generate faces between adjacent cells (PARALLEL)
    // Create a face when exactly one of two neighboring cells has a vertex
    let face_coords: Vec<(u32, u32, u32)> = cell_vertices.keys().copied().collect();

    let indices = Mutex::new(Vec::new());

    face_coords.par_iter().for_each(|&(x, y, z)| {
        let mut local_indices = Vec::new();

        // Create face in +X direction (YZ plane)
        // Needs cells: (x,y,z), (x,y,z+1), (x,y+1,z+1), (x,y+1,z)
        if y < res - 2 && z < res - 2 {
            create_face_x(&cell_vertices, &mut local_indices, x, y, z);
        }

        // Create face in +Y direction (XZ plane)
        // Needs cells: (x,y,z), (x+1,y,z), (x+1,y,z+1), (x,y,z+1)
        if x < res - 2 && z < res - 2 {
            create_face_y(&cell_vertices, &mut local_indices, x, y, z);
        }

        // Create face in +Z direction (XY plane)
        // Needs cells: (x,y,z), (x+1,y,z), (x+1,y+1,z), (x,y+1,z)
        if x < res - 2 && y < res - 2 {
            create_face_z(&cell_vertices, &mut local_indices, x, y, z);
        }

        if !local_indices.is_empty() {
            indices.lock().unwrap().extend(local_indices);
        }
    });

    let indices = indices.into_inner().unwrap();

    // Compute normals (simple per-face normals for now)
    let normals = compute_normals(&vertices, &indices);

    MeshData {
        vertices,
        indices,
        normals,
    }
}

/// Check if a voxel cell intersects the isosurface
/// Returns true if corners have different signs (sign change = surface crossing)
fn cell_intersects_surface(grid: &VoxelGrid, x: u32, y: u32, z: u32, iso_value: f32) -> bool {
    // Get 8 corner values
    let corners = [
        grid.get(x, y, z),
        grid.get(x + 1, y, z),
        grid.get(x + 1, y, z + 1),
        grid.get(x, y, z + 1),
        grid.get(x, y + 1, z),
        grid.get(x + 1, y + 1, z),
        grid.get(x + 1, y + 1, z + 1),
        grid.get(x, y + 1, z + 1),
    ];

    // Check if any corner is inside (< isoValue) and any is outside (>= isoValue)
    let mut has_inside = false;
    let mut has_outside = false;

    for &value in &corners {
        if value < iso_value {
            has_inside = true;
        } else {
            has_outside = true;
        }
    }

    has_inside && has_outside
}

/// Find optimal vertex position for a cell using surface projection
/// Uses Newton's method to find the closest point on the isosurface to the cell center
fn find_cell_vertex(
    grid: &VoxelGrid,
    mould_manager: &MouldManager,
    x: u32,
    y: u32,
    z: u32,
    iso_value: f32,
) -> Pt3 {
    // Start at cell center
    let cell_center = grid.get_position(x as f32 + 0.5, y as f32 + 0.5, z as f32 + 0.5);
    let mut pos = cell_center;

    // Use Newton's method for fast convergence to isosurface
    // Reduced iterations for better performance (was 20, now 8)
    let max_iterations = 8;
    let tolerance = 0.001; // Slightly relaxed tolerance

    for _ in 0..max_iterations {
        let dist = mould_manager.evaluate_sdf(&pos) - iso_value;

        // Close enough to surface
        if dist.abs() < tolerance {
            break;
        }

        // Compute gradient at current position
        let grad = compute_gradient(&pos, |p| mould_manager.evaluate_sdf(p));

        // Gradient magnitude for normalization
        let grad_len = grad.magnitude();

        if grad_len < 0.0001 {
            // Gradient too small, can't make progress
            break;
        }

        // Newton's method: move exactly to the surface along gradient direction
        // Distance to surface divided by gradient magnitude gives step size
        let step_size = dist / grad_len;

        pos = Pt3::new(
            pos.x - grad.x * step_size,
            pos.y - grad.y * step_size,
            pos.z - grad.z * step_size,
        );
    }

    pos
}

/// Create face between cells in X direction (perpendicular to X-axis)
/// Connects 4 cells in a 2x2 grid in the YZ plane
/// Winding order determined by face normal pointing outward
fn create_face_x(
    cell_vertices: &HashMap<(u32, u32, u32), CellVertex>,
    indices: &mut Vec<u32>,
    x: u32,
    y: u32,
    z: u32,
) {
    // Four cells form a quad in the YZ plane perpendicular to X-axis
    let v0 = cell_vertices.get(&(x, y, z));
    let v1 = cell_vertices.get(&(x, y, z + 1));
    let v2 = cell_vertices.get(&(x, y + 1, z + 1));
    let v3 = cell_vertices.get(&(x, y + 1, z));

    if v0.is_none() || v1.is_none() || v2.is_none() || v3.is_none() {
        return;
    }

    let v0 = v0.unwrap();
    let v1 = v1.unwrap();
    let v2 = v2.unwrap();
    let v3 = v3.unwrap();

    // Compute face normal using cross product
    // Edge from v0 to v1, edge from v0 to v3
    let e1 = v1.position - v0.position;
    let e2 = v3.position - v0.position;
    let face_normal = e1.cross(&e2);

    // Check if face normal points in +X or -X direction
    // If normal.x > 0, we want CCW from +X view
    // If normal.x < 0, we want CW from +X view (which is CCW from -X view)
    let flip = face_normal.x < 0.0;

    // Triangulate quad along shortest diagonal
    let diag02 = distance(&v0.position, &v2.position);
    let diag13 = distance(&v1.position, &v3.position);

    if diag02 < diag13 {
        // Diagonal from v0 to v2
        if flip {
            indices.push(v0.index);
            indices.push(v2.index);
            indices.push(v1.index);
            indices.push(v0.index);
            indices.push(v3.index);
            indices.push(v2.index);
        } else {
            indices.push(v0.index);
            indices.push(v1.index);
            indices.push(v2.index);
            indices.push(v0.index);
            indices.push(v2.index);
            indices.push(v3.index);
        }
    } else {
        // Diagonal from v1 to v3
        if flip {
            indices.push(v0.index);
            indices.push(v3.index);
            indices.push(v1.index);
            indices.push(v1.index);
            indices.push(v3.index);
            indices.push(v2.index);
        } else {
            indices.push(v0.index);
            indices.push(v1.index);
            indices.push(v3.index);
            indices.push(v1.index);
            indices.push(v2.index);
            indices.push(v3.index);
        }
    }
}

/// Create face between cells in Y direction (perpendicular to Y-axis)
/// Connects 4 cells in a 2x2 grid in the XZ plane
/// This face is created when the Y-axis edge crosses the isosurface
/// Winding order is CCW when viewed from +Y direction (outside)
fn create_face_y(
    cell_vertices: &HashMap<(u32, u32, u32), CellVertex>,
    indices: &mut Vec<u32>,
    x: u32,
    y: u32,
    z: u32,
) {
    // Four cells form a quad in the XZ plane perpendicular to Y-axis
    // The quad surrounds the Y-axis edge at (x,y,z) going to (x,y+1,z)
    let v0 = cell_vertices.get(&(x, y, z));         // (-X, -Z) relative to edge
    let v1 = cell_vertices.get(&(x + 1, y, z));     // (+X, -Z)
    let v2 = cell_vertices.get(&(x + 1, y, z + 1)); // (+X, +Z)
    let v3 = cell_vertices.get(&(x, y, z + 1));     // (-X, +Z)

    if v0.is_none() || v1.is_none() || v2.is_none() || v3.is_none() {
        return;
    }

    let v0 = v0.unwrap();
    let v1 = v1.unwrap();
    let v2 = v2.unwrap();
    let v3 = v3.unwrap();

    let diag02 = distance(&v0.position, &v2.position);
    let diag13 = distance(&v1.position, &v3.position);

    if diag02 < diag13 {
        // Diagonal from v0 to v2
        indices.push(v0.index);
        indices.push(v1.index);
        indices.push(v2.index);

        indices.push(v0.index);
        indices.push(v2.index);
        indices.push(v3.index);
    } else {
        // Diagonal from v1 to v3
        indices.push(v0.index);
        indices.push(v1.index);
        indices.push(v3.index);

        indices.push(v1.index);
        indices.push(v2.index);
        indices.push(v3.index);
    }
}

/// Create face between cells in Z direction (perpendicular to Z-axis)
/// Connects 4 cells in a 2x2 grid in the XY plane
/// This face is created when the Z-axis edge crosses the isosurface
/// Winding order is CCW when viewed from +Z direction (outside)
fn create_face_z(
    cell_vertices: &HashMap<(u32, u32, u32), CellVertex>,
    indices: &mut Vec<u32>,
    x: u32,
    y: u32,
    z: u32,
) {
    // Four cells form a quad in the XY plane perpendicular to Z-axis
    // The quad surrounds the Z-axis edge at (x,y,z) going to (x,y,z+1)
    let v0 = cell_vertices.get(&(x, y, z));         // (-X, -Y) relative to edge
    let v1 = cell_vertices.get(&(x + 1, y, z));     // (+X, -Y)
    let v2 = cell_vertices.get(&(x + 1, y + 1, z)); // (+X, +Y)
    let v3 = cell_vertices.get(&(x, y + 1, z));     // (-X, +Y)

    if v0.is_none() || v1.is_none() || v2.is_none() || v3.is_none() {
        return;
    }

    let v0 = v0.unwrap();
    let v1 = v1.unwrap();
    let v2 = v2.unwrap();
    let v3 = v3.unwrap();

    let diag02 = distance(&v0.position, &v2.position);
    let diag13 = distance(&v1.position, &v3.position);

    if diag02 < diag13 {
        // Diagonal from v0 to v2
        indices.push(v0.index);
        indices.push(v1.index);
        indices.push(v2.index);

        indices.push(v0.index);
        indices.push(v2.index);
        indices.push(v3.index);
    } else {
        // Diagonal from v1 to v3
        indices.push(v0.index);
        indices.push(v1.index);
        indices.push(v3.index);

        indices.push(v1.index);
        indices.push(v2.index);
        indices.push(v3.index);
    }
}

/// Compute Euclidean distance between two points
fn distance(a: &Pt3, b: &Pt3) -> f32 {
    (a - b).magnitude()
}

/// Compute per-vertex normals from triangle mesh
fn compute_normals(vertices: &[f32], indices: &[u32]) -> Vec<f32> {
    let num_vertices = vertices.len() / 3;
    let mut normals = vec![0.0; vertices.len()];

    // Accumulate face normals for each vertex
    for i in (0..indices.len()).step_by(3) {
        let i0 = indices[i] as usize;
        let i1 = indices[i + 1] as usize;
        let i2 = indices[i + 2] as usize;

        // Get triangle vertices
        let v0 = Vec3::new(
            vertices[i0 * 3],
            vertices[i0 * 3 + 1],
            vertices[i0 * 3 + 2],
        );
        let v1 = Vec3::new(
            vertices[i1 * 3],
            vertices[i1 * 3 + 1],
            vertices[i1 * 3 + 2],
        );
        let v2 = Vec3::new(
            vertices[i2 * 3],
            vertices[i2 * 3 + 1],
            vertices[i2 * 3 + 2],
        );

        // Compute face normal
        let edge1 = v1 - v0;
        let edge2 = v2 - v0;
        let normal = edge1.cross(&edge2);

        // Accumulate to vertex normals
        for &idx in &[i0, i1, i2] {
            normals[idx * 3] += normal.x;
            normals[idx * 3 + 1] += normal.y;
            normals[idx * 3 + 2] += normal.z;
        }
    }

    // Normalize all vertex normals
    for i in 0..num_vertices {
        let nx = normals[i * 3];
        let ny = normals[i * 3 + 1];
        let nz = normals[i * 3 + 2];
        let len = (nx * nx + ny * ny + nz * nz).sqrt();

        if len > 0.0001 {
            normals[i * 3] /= len;
            normals[i * 3 + 1] /= len;
            normals[i * 3 + 2] /= len;
        }
    }

    normals
}

/// Dual contouring for BrickMap (high-resolution sparse grids)
pub fn dual_contouring_brick_map(
    brick_map: &BrickMap,
    mould_manager: &MouldManager,
    iso_value: f32,
    fast_mode: bool,
) -> MeshData {
    // Use the generic Grid trait implementation
    dual_contouring_generic(brick_map, mould_manager, iso_value, fast_mode)
}

/// Generic dual contouring that works with any Grid implementation
fn dual_contouring_generic<G: Grid + Sync>(
    grid: &G,
    mould_manager: &MouldManager,
    iso_value: f32,
    fast_mode: bool,
) -> MeshData {
    let res = grid.resolution();

    // Step 1: Create vertices for cells that intersect the isosurface (PARALLEL)
    // Generate all cell coordinates
    let cell_coords: Vec<(u32, u32, u32)> = (0..res - 1)
        .flat_map(|z| {
            (0..res - 1).flat_map(move |y| {
                (0..res - 1).map(move |x| (x, y, z))
            })
        })
        .collect();

    // Parallel vertex creation
    let surface_cells: Vec<((u32, u32, u32), Pt3)> = cell_coords
        .par_iter()
        .filter_map(|&(x, y, z)| {
            // Check if this cell intersects the isosurface
            if !cell_intersects_surface_generic(grid, x, y, z, iso_value) {
                return None;
            }

            // Find best vertex position for this cell
            let vertex_pos = if fast_mode {
                // Fast mode: just use cell center (no Newton iteration)
                grid.get_position(x as f32 + 0.5, y as f32 + 0.5, z as f32 + 0.5)
            } else {
                find_cell_vertex_generic(grid, mould_manager, x, y, z, iso_value)
            };

            Some(((x, y, z), vertex_pos))
        })
        .collect();

    // Build vertices array and cell_vertices map
    let mut vertices: Vec<f32> = Vec::with_capacity(surface_cells.len() * 3);
    let mut cell_vertices: HashMap<(u32, u32, u32), CellVertex> = HashMap::with_capacity(surface_cells.len());

    for ((x, y, z), vertex_pos) in surface_cells {
        let vertex_index = (vertices.len() / 3) as u32;
        vertices.push(vertex_pos.x);
        vertices.push(vertex_pos.y);
        vertices.push(vertex_pos.z);

        cell_vertices.insert(
            (x, y, z),
            CellVertex {
                position: vertex_pos,
                index: vertex_index,
            },
        );
    }

    // Step 2: Generate faces between adjacent cells (PARALLEL)
    let face_coords: Vec<(u32, u32, u32)> = cell_vertices.keys().copied().collect();

    let indices = Mutex::new(Vec::new());

    face_coords.par_iter().for_each(|&(x, y, z)| {
        let mut local_indices = Vec::new();

        // Create face in +X direction (YZ plane)
        if y < res - 2 && z < res - 2 {
            create_face_x(&cell_vertices, &mut local_indices, x, y, z);
        }

        // Create face in +Y direction (XZ plane)
        if x < res - 2 && z < res - 2 {
            create_face_y(&cell_vertices, &mut local_indices, x, y, z);
        }

        // Create face in +Z direction (XY plane)
        if x < res - 2 && y < res - 2 {
            create_face_z(&cell_vertices, &mut local_indices, x, y, z);
        }

        if !local_indices.is_empty() {
            indices.lock().unwrap().extend(local_indices);
        }
    });

    let indices = indices.into_inner().unwrap();

    // Compute normals
    let normals = compute_normals(&vertices, &indices);

    MeshData {
        vertices,
        indices,
        normals,
    }
}

/// Check if a voxel cell intersects the isosurface (generic version)
fn cell_intersects_surface_generic<G: Grid>(
    grid: &G,
    x: u32,
    y: u32,
    z: u32,
    iso_value: f32,
) -> bool {
    // Get 8 corner values
    let corners = [
        grid.get(x, y, z),
        grid.get(x + 1, y, z),
        grid.get(x + 1, y, z + 1),
        grid.get(x, y, z + 1),
        grid.get(x, y + 1, z),
        grid.get(x + 1, y + 1, z),
        grid.get(x + 1, y + 1, z + 1),
        grid.get(x, y + 1, z + 1),
    ];

    // Check if any corner is inside (< isoValue) and any is outside (>= isoValue)
    let mut has_inside = false;
    let mut has_outside = false;

    for &value in &corners {
        if value < iso_value {
            has_inside = true;
        } else {
            has_outside = true;
        }
    }

    has_inside && has_outside
}

/// Find optimal vertex position for a cell (generic version)
fn find_cell_vertex_generic<G: Grid>(
    grid: &G,
    mould_manager: &MouldManager,
    x: u32,
    y: u32,
    z: u32,
    iso_value: f32,
) -> Pt3 {
    // Start at cell center
    let cell_center = grid.get_position(x as f32 + 0.5, y as f32 + 0.5, z as f32 + 0.5);
    let mut pos = cell_center;

    // Use Newton's method for fast convergence to isosurface
    let max_iterations = 8;
    let tolerance = 0.001;

    for _ in 0..max_iterations {
        let dist = mould_manager.evaluate_sdf(&pos) - iso_value;

        // Close enough to surface
        if dist.abs() < tolerance {
            break;
        }

        // Compute gradient at current position
        let grad = compute_gradient(&pos, |p| mould_manager.evaluate_sdf(p));

        // Gradient magnitude for normalization
        let grad_len = grad.magnitude();

        if grad_len < 0.0001 {
            // Gradient too small, can't make progress
            break;
        }

        // Newton's method: move exactly to the surface along gradient direction
        let step_size = dist / grad_len;

        pos = Pt3::new(
            pos.x - grad.x * step_size,
            pos.y - grad.y * step_size,
            pos.z - grad.z * step_size,
        );
    }

    pos
}
